@BsonDiscriminator not fully working

Although it's not documented in Micronaut Mongo, a feature of the mongo driver is that you can map elements of a collection to specific subclasses based on a field called the discriminator, see this example in C# https://stackoverflow.com/questions/18465529/mongodb-mapping-an-interface-to-an-implementation
Or the docs for this feature in the C# driver https://mongodb.github.io/mongo-csharp-driver/2.8/reference/bson/mapping/polymorphism/

My issue is that I could not find a mechanism to register classes with the automatic Pojo codec, so the code is flaky. For some reason at insert time the inserted class is registered automatically, so my current workaround for this is to insert and delete one item of every type I'm dealing with. The mapping of discriminator key to Class is in org.bson.codecs.pojo.DiscriminatorLookup

See this example: https://github.com/mnembrini/micronaut-bug-bson-discriminator

In there there is an Animal base class with subclasses Dog and Cat, and a single mongoDB collection "animal". The expectation is that we get the right class when reading based on the discriminator key "type"

In AnimalRepositoryTest, if you run the whole test class, everything passes, because there is an insert of both Dog and Cat classes in the DB.
If after that you run only the test readCatWithoutInsertingFirst(), it will fail with the following exception

```
An exception occurred when decoding using the AutomaticPojoCodec.
Decoding into a 'Animal' failed with the following exception:

Failed to decode 'Animal'. Decoding errored with: A class could not be found for the discriminator: 'some_cat_type'.

A custom Codec or PojoCodec may need to be explicitly configured and registered to handle this type.
org.bson.codecs.configuration.CodecConfigurationException: An exception occurred when decoding using the AutomaticPojoCodec.
Decoding into a 'Animal' failed with the following exception:

Failed to decode 'Animal'. Decoding errored with: A class could not be found for the discriminator: 'some_cat_type'.

A custom Codec or PojoCodec may need to be explicitly configured and registered to handle this type.
	at org.bson.codecs.pojo.AutomaticPojoCodec.decode(AutomaticPojoCodec.java:40)
	at com.mongodb.operation.CommandResultArrayCodec.decode(CommandResultArrayCodec.java:52)
	at com.mongodb.operation.CommandResultDocumentCodec.readValue(CommandResultDocumentCodec.java:60)
	at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:84)
	at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:41)
	at org.bson.codecs.configuration.LazyCodec.decode(LazyCodec.java:47)
	at org.bson.codecs.BsonDocumentCodec.readValue(BsonDocumentCodec.java:101)
	at com.mongodb.operation.CommandResultDocumentCodec.readValue(CommandResultDocumentCodec.java:63)
	at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:84)
	at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:41)
	at com.mongodb.internal.connection.ReplyMessage.<init>(ReplyMessage.java:51)
	at com.mongodb.internal.connection.InternalStreamConnection.getCommandResult(InternalStreamConnection.java:409)
	at com.mongodb.internal.connection.InternalStreamConnection.access$900(InternalStreamConnection.java:74)
	at com.mongodb.internal.connection.InternalStreamConnection$2$1.onResult(InternalStreamConnection.java:394)
	at com.mongodb.internal.connection.InternalStreamConnection$2$1.onResult(InternalStreamConnection.java:372)
	at com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback$MessageCallback.onResult(InternalStreamConnection.java:667)
	at com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback$MessageCallback.onResult(InternalStreamConnection.java:634)
	at com.mongodb.internal.connection.InternalStreamConnection$5.completed(InternalStreamConnection.java:510)
	at com.mongodb.internal.connection.InternalStreamConnection$5.completed(InternalStreamConnection.java:507)
	at com.mongodb.internal.connection.AsynchronousChannelStream$BasicCompletionHandler.completed(AsynchronousChannelStream.java:220)
	at com.mongodb.internal.connection.AsynchronousChannelStream$BasicCompletionHandler.completed(AsynchronousChannelStream.java:203)
	at java.base/sun.nio.ch.Invoker.invokeUnchecked(Invoker.java:127)
	at java.base/sun.nio.ch.Invoker.invokeDirect(Invoker.java:158)
	at java.base/sun.nio.ch.UnixAsynchronousSocketChannelImpl.implRead(UnixAsynchronousSocketChannelImpl.java:562)
	at java.base/sun.nio.ch.AsynchronousSocketChannelImpl.read(AsynchronousSocketChannelImpl.java:277)
	at java.base/sun.nio.ch.AsynchronousSocketChannelImpl.read(AsynchronousSocketChannelImpl.java:298)
	at com.mongodb.internal.connection.AsynchronousSocketChannelStream$AsynchronousSocketChannelAdapter.read(AsynchronousSocketChannelStream.java:137)
	at com.mongodb.internal.connection.AsynchronousChannelStream.readAsync(AsynchronousChannelStream.java:105)
	at com.mongodb.internal.connection.InternalStreamConnection.readAsync(InternalStreamConnection.java:507)
	at com.mongodb.internal.connection.InternalStreamConnection.access$1000(InternalStreamConnection.java:74)
	at com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback.onResult(InternalStreamConnection.java:624)
	at com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback.onResult(InternalStreamConnection.java:609)
	at com.mongodb.internal.connection.InternalStreamConnection$5.completed(InternalStreamConnection.java:510)
	at com.mongodb.internal.connection.InternalStreamConnection$5.completed(InternalStreamConnection.java:507)
	at com.mongodb.internal.connection.AsynchronousChannelStream$BasicCompletionHandler.completed(AsynchronousChannelStream.java:220)
	at com.mongodb.internal.connection.AsynchronousChannelStream$BasicCompletionHandler.completed(AsynchronousChannelStream.java:203)
	at java.base/sun.nio.ch.Invoker.invokeUnchecked(Invoker.java:127)
	at java.base/sun.nio.ch.Invoker$2.run(Invoker.java:219)
	at java.base/sun.nio.ch.AsynchronousChannelGroupImpl$1.run(AsynchronousChannelGroupImpl.java:112)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:835)
Caused by: org.bson.codecs.configuration.CodecConfigurationException: Failed to decode 'Animal'. Decoding errored with: A class could not be found for the discriminator: 'some_cat_type'.
	at org.bson.codecs.pojo.PojoCodecImpl.getCodecFromDocument(PojoCodecImpl.java:354)
	at org.bson.codecs.pojo.PojoCodecImpl.decode(PojoCodecImpl.java:124)
	at org.bson.codecs.pojo.AutomaticPojoCodec.decode(AutomaticPojoCodec.java:37)
	... 41 more
Caused by: org.bson.codecs.configuration.CodecConfigurationException: A class could not be found for the discriminator: 'some_cat_type'.
	at org.bson.codecs.pojo.DiscriminatorLookup.lookup(DiscriminatorLookup.java:51)
	at org.bson.codecs.pojo.PojoCodecImpl.getCodecFromDocument(PojoCodecImpl.java:352)
	... 43 more
```

The test readDogWithoutInsertingFirst instead works, because the discriminator key for Dog is the name of the class. This is also another possible workaround, but I don't think it's robust

--- a/mongo-core/src/main/java/io/micronaut/configuration/mongo/core/DefaultCodecRegistryBuilder.java
+++ b/mongo-core/src/main/java/io/micronaut/configuration/mongo/core/DefaultCodecRegistryBuilder.java
@@ -34,6 +34,9 @@ import org.bson.codecs.pojo.annotations.BsonDiscriminator;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
@@ -54,6 +57,7 @@ public final class DefaultCodecRegistryBuilder implements CodecRegistryBuilder {
 
     private final Environment environment;
     private final BeanProvider<SerdeRegistry> serdeRegistry;
+    private final Map<List<String>, Collection<Class<?>>> discriminatorEntitiesCache = new ConcurrentHashMap<>();
 
     public DefaultCodecRegistryBuilder(Environment environment, BeanProvider<SerdeRegistry> serdeRegistry) {
         this.environment = environment;
@@ -100,10 +104,17 @@ public final class DefaultCodecRegistryBuilder implements CodecRegistryBuilder {
     }
 
     private Collection<Class<?>> findDiscriminatorEntities(Collection<String> packageNames) {
-        return BeanIntrospector.SHARED.findIntrospectedTypes(reference ->
-            reference.isPresent()
-                && reference.isAnnotationPresent(BsonDiscriminator.class)
-                && isWithinConfiguredPackages(reference.getBeanType(), packageNames)
+        List<String> packageNamesKey = packageNames.stream()
+            .filter(Objects::nonNull)
+            .distinct()
+            .sorted()
+            .toList();
+        return discriminatorEntitiesCache.computeIfAbsent(packageNamesKey, key ->
+            List.copyOf(BeanIntrospector.SHARED.findIntrospectedTypes(reference ->
+                reference.isPresent()
+                    && reference.isAnnotationPresent(BsonDiscriminator.class)
+                    && isWithinConfiguredPackages(reference.getBeanType(), key)
+            ))
         );
     }
 

--- a/mongo-core/src/main/java/io/micronaut/configuration/mongo/core/DefaultCodecRegistryBuilder.java
+++ b/mongo-core/src/main/java/io/micronaut/configuration/mongo/core/DefaultCodecRegistryBuilder.java
@@ -104,18 +104,27 @@ public final class DefaultCodecRegistryBuilder implements CodecRegistryBuilder {
     }
 
     private Collection<Class<?>> findDiscriminatorEntities(Collection<String> packageNames) {
-        List<String> packageNamesKey = packageNames.stream()
+        List<String> packageNamesKey = normalizePackageNamesForCaching(packageNames);
+        Collection<Class<?>> cachedEntities = discriminatorEntitiesCache.get(packageNamesKey);
+        if (cachedEntities != null) {
+            return cachedEntities;
+        }
+        Collection<Class<?>> discoveredEntities = List.copyOf(BeanIntrospector.SHARED.findIntrospectedTypes(reference ->
+            reference.isPresent()
+                && reference.isAnnotationPresent(BsonDiscriminator.class)
+                && isWithinConfiguredPackages(reference.getBeanType(), packageNamesKey)
+        ));
+        Collection<Class<?>> existingEntities = discriminatorEntitiesCache.putIfAbsent(packageNamesKey, discoveredEntities);
+        return existingEntities != null ? existingEntities : discoveredEntities;
+    }
+
+    private List<String> normalizePackageNamesForCaching(Collection<String> packageNames) {
+        // Normalize order/duplicates so logically equivalent package sets share a single cache key.
+        return packageNames.stream()
             .filter(Objects::nonNull)
             .distinct()
             .sorted()
             .toList();
-        return discriminatorEntitiesCache.computeIfAbsent(packageNamesKey, key ->
-            List.copyOf(BeanIntrospector.SHARED.findIntrospectedTypes(reference ->
-                reference.isPresent()
-                    && reference.isAnnotationPresent(BsonDiscriminator.class)
-                    && isWithinConfiguredPackages(reference.getBeanType(), key)
-            ))
-        );
     }
 
     private boolean isWithinConfiguredPackages(Class<?> beanType, Collection<String> packageNames) {

--- a/mongo-core/src/main/java/io/micronaut/configuration/mongo/core/DefaultCodecRegistryBuilder.java
+++ b/mongo-core/src/main/java/io/micronaut/configuration/mongo/core/DefaultCodecRegistryBuilder.java
@@ -22,12 +22,14 @@ import io.micronaut.context.annotation.Prototype;
 import io.micronaut.context.annotation.Requires;
 import io.micronaut.context.env.Environment;
 import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.beans.BeanIntrospector;
 import io.micronaut.core.util.CollectionUtils;
 import io.micronaut.serde.SerdeRegistry;
 import io.micronaut.serde.annotation.Serdeable;
 import org.bson.codecs.Codec;
 import org.bson.codecs.configuration.CodecRegistry;
 import org.bson.codecs.pojo.PojoCodecProvider;
+import org.bson.codecs.pojo.annotations.BsonDiscriminator;
 
 import java.util.ArrayList;
 import java.util.Collection;
@@ -84,14 +86,31 @@ public final class DefaultCodecRegistryBuilder implements CodecRegistryBuilder {
         } else {
             final PojoCodecProvider.Builder builder = PojoCodecProvider.builder();
             if (CollectionUtils.isNotEmpty(packageNames)) {
+                Collection<Class<?>> discriminatorEntities = findDiscriminatorEntities(packageNames);
+                if (CollectionUtils.isNotEmpty(discriminatorEntities)) {
+                    builder.register(discriminatorEntities.toArray(Class<?>[]::new));
+                }
                 builder.register(packageNames.toArray(new String[0]));
             }
-            codecRegistries.add(
-                fromProviders(
-                    builder.automatic(configuration.isAutomaticClassModels()).build()
-                )
-            );
+            codecRegistries.add(fromProviders(
+                builder.automatic(configuration.isAutomaticClassModels()).build()
+            ));
         }
         return fromRegistries(codecRegistries);
+    }
+
+    private Collection<Class<?>> findDiscriminatorEntities(Collection<String> packageNames) {
+        return BeanIntrospector.SHARED.findIntrospectedTypes(reference ->
+            reference.isPresent()
+                && reference.isAnnotationPresent(BsonDiscriminator.class)
+                && isWithinConfiguredPackages(reference.getBeanType(), packageNames)
+        );
+    }
+
+    private boolean isWithinConfiguredPackages(Class<?> beanType, Collection<String> packageNames) {
+        String packageName = beanType.getPackageName();
+        return packageNames.stream().anyMatch(configuredPackage ->
+            packageName.equals(configuredPackage) || packageName.startsWith(configuredPackage + ".")
+        );
     }
 }

--- a/src/main/docs/guide/config.adoc
+++ b/src/main/docs/guide/config.adoc
@@ -28,6 +28,20 @@ mongodb:
         maxSize: 20
 ----
 
+==== POJO Discriminators
+
+When using MongoDB POJO codecs together with `@BsonDiscriminator`, configure `mongodb.package-names` so Micronaut can register the matching types with the MongoDB `PojoCodecProvider`.
+
+To avoid classpath or JAR scanning, discriminator registration only considers classes that have Micronaut bean introspections. Annotate the root type and subtypes with `@Introspected`, or otherwise provide bean introspection metadata for those classes.
+
+.Configuring packages for discriminator-based POJOs
+[configuration]
+----
+mongodb:
+  package-names:
+    - example.animals
+----
+
 ==== Multiple MongoDB Drivers
 
 You can create multiple MongoDB connections using the `mongodb.servers` setting. For example in `application.yml`:

--- a/tests/mongo-pojo/src/test/java/example/Animal.java
+++ b/tests/mongo-pojo/src/test/java/example/Animal.java
@@ -1,0 +1,21 @@
+package example;
+
+import io.micronaut.core.annotation.Introspected;
+import org.bson.codecs.pojo.annotations.BsonDiscriminator;
+
+@Introspected
+@BsonDiscriminator(key = "type")
+public class Animal {
+    private String name;
+
+    public Animal() {
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+}

--- a/tests/mongo-pojo/src/test/java/example/BsonDiscriminatorTest.java
+++ b/tests/mongo-pojo/src/test/java/example/BsonDiscriminatorTest.java
@@ -1,0 +1,58 @@
+package example;
+
+import io.micronaut.configuration.mongo.core.DefaultCodecRegistryBuilder;
+import io.micronaut.configuration.mongo.core.DefaultMongoConfiguration;
+import io.micronaut.core.beans.BeanIntrospector;
+import io.micronaut.context.ApplicationContext;
+import io.micronaut.runtime.ApplicationConfiguration;
+import org.bson.BsonDocument;
+import org.bson.BsonDocumentReader;
+import org.bson.BsonString;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.codecs.pojo.annotations.BsonDiscriminator;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+class BsonDiscriminatorTest {
+
+    @Test
+    void decodesCustomDiscriminatorWithoutPrimingSubtypeCodec() {
+        try (ApplicationContext context = ApplicationContext.run()) {
+            DefaultMongoConfiguration configuration = new DefaultMongoConfiguration(
+                new ApplicationConfiguration(),
+                context.getEnvironment()
+            );
+            configuration.setPackageNames(List.of("example"));
+
+            Set<Class<?>> discriminatorEntities = BeanIntrospector.SHARED.findIntrospectedTypes(reference ->
+                    reference.isAnnotationPresent(BsonDiscriminator.class)
+                        && reference.getBeanType().getPackageName().equals("example")
+                ).stream()
+                .collect(Collectors.toSet());
+            Assertions.assertTrue(discriminatorEntities.contains(Animal.class));
+            Assertions.assertTrue(discriminatorEntities.contains(Dog.class));
+
+            CodecRegistry codecRegistry = new DefaultCodecRegistryBuilder(
+                context.getEnvironment(),
+                null
+            ).build(configuration);
+
+            Animal animal = codecRegistry.get(Animal.class).decode(
+                new BsonDocumentReader(new BsonDocument()
+                    .append("type", new BsonString("dog"))
+                    .append("name", new BsonString("Fido"))
+                    .append("favoriteToy", new BsonString("Ball"))),
+                DecoderContext.builder().build()
+            );
+
+            Assertions.assertInstanceOf(Dog.class, animal);
+            Assertions.assertEquals("Fido", animal.getName());
+            Assertions.assertEquals("Ball", ((Dog) animal).getFavoriteToy());
+        }
+    }
+}

--- a/tests/mongo-pojo/src/test/java/example/BsonDiscriminatorTest.java
+++ b/tests/mongo-pojo/src/test/java/example/BsonDiscriminatorTest.java
@@ -19,6 +19,7 @@ import java.util.Set;
 import java.util.stream.Collectors;
 
 class BsonDiscriminatorTest {
+    private static final String TEST_PACKAGE = "example";
 
     @Test
     void decodesCustomDiscriminatorWithoutPrimingSubtypeCodec() {
@@ -27,13 +28,12 @@ class BsonDiscriminatorTest {
                 new ApplicationConfiguration(),
                 context.getEnvironment()
             );
-            String configuredPackage = "example";
-            configuration.setPackageNames(List.of(configuredPackage));
+            configuration.setPackageNames(List.of(TEST_PACKAGE));
 
             Set<Class<?>> discriminatorEntities = BeanIntrospector.SHARED.findIntrospectedTypes(reference -> {
                     String packageName = reference.getBeanType().getPackageName();
                     return reference.isAnnotationPresent(BsonDiscriminator.class)
-                        && (packageName.equals(configuredPackage) || packageName.startsWith(configuredPackage + "."));
+                        && (packageName.equals(TEST_PACKAGE) || packageName.startsWith(TEST_PACKAGE + "."));
                 }).stream()
                 .collect(Collectors.toSet());
             Assertions.assertTrue(discriminatorEntities.contains(Animal.class));

--- a/tests/mongo-pojo/src/test/java/example/BsonDiscriminatorTest.java
+++ b/tests/mongo-pojo/src/test/java/example/BsonDiscriminatorTest.java
@@ -27,20 +27,19 @@ class BsonDiscriminatorTest {
                 new ApplicationConfiguration(),
                 context.getEnvironment()
             );
-            configuration.setPackageNames(List.of("example"));
+            String configuredPackage = "example";
+            configuration.setPackageNames(List.of(configuredPackage));
 
-            Set<Class<?>> discriminatorEntities = BeanIntrospector.SHARED.findIntrospectedTypes(reference ->
-                    reference.isAnnotationPresent(BsonDiscriminator.class)
-                        && reference.getBeanType().getPackageName().equals("example")
-                ).stream()
+            Set<Class<?>> discriminatorEntities = BeanIntrospector.SHARED.findIntrospectedTypes(reference -> {
+                    String packageName = reference.getBeanType().getPackageName();
+                    return reference.isAnnotationPresent(BsonDiscriminator.class)
+                        && (packageName.equals(configuredPackage) || packageName.startsWith(configuredPackage + "."));
+                }).stream()
                 .collect(Collectors.toSet());
             Assertions.assertTrue(discriminatorEntities.contains(Animal.class));
             Assertions.assertTrue(discriminatorEntities.contains(Dog.class));
 
-            CodecRegistry codecRegistry = new DefaultCodecRegistryBuilder(
-                context.getEnvironment(),
-                null
-            ).build(configuration);
+            CodecRegistry codecRegistry = context.getBean(DefaultCodecRegistryBuilder.class).build(configuration);
 
             Animal animal = codecRegistry.get(Animal.class).decode(
                 new BsonDocumentReader(new BsonDocument()

--- a/tests/mongo-pojo/src/test/java/example/Dog.java
+++ b/tests/mongo-pojo/src/test/java/example/Dog.java
@@ -1,0 +1,21 @@
+package example;
+
+import io.micronaut.core.annotation.Introspected;
+import org.bson.codecs.pojo.annotations.BsonDiscriminator;
+
+@Introspected
+@BsonDiscriminator(key = "type", value = "dog")
+public class Dog extends Animal {
+    private String favoriteToy;
+
+    public Dog() {
+    }
+
+    public String getFavoriteToy() {
+        return favoriteToy;
+    }
+
+    public void setFavoriteToy(String favoriteToy) {
+        this.favoriteToy = favoriteToy;
+    }
+}

